chia/full_node/weight_proof_v2.py

```diff
@@ -0,0 +1,1758 @@
+    C = 0.5
+    MAX_SAMPLES = 140
+
+    def __init__(
+        self,
+        constants: ConsensusConstants,
+        blockchain: BlockchainInterface,
+    ):
+        self.tip: Optional[bytes32] = None
+        self.proof: Optional[WeightProofV2] = None
+        self.constants = constants
+        self.blockchain = blockchain
+        self.lock = asyncio.Lock()
+
+    async def get_proof_of_weight(self, tip: bytes32, seed: bytes32) -> Optional[WeightProofV2]:
+
+        tip_rec = self.blockchain.try_block_record(tip)
+        if tip_rec is None:
+            log.error("unknown tip")
+            return None
+
+        if tip_rec.height < self.constants.WEIGHT_PROOF_RECENT_BLOCKS:
+            log.debug("need at least 3 sub epochs for weight proof")
+            return None
+
+        async with self.lock:
+            wp = await self._create_proof_of_weight(tip, seed)
+            if wp is None:
+                return None
+            self.proof = wp
+            self.tip = tip
+            return wp
+
+    def get_fork_point_no_validations(self, weight_proof: WeightProofV2) -> Tuple[bool, uint32]:
+        log.debug("get fork point skip validations")
+        if len(weight_proof.sub_epochs) == 0:
+            return False, uint32(0)
+        summaries, sub_epoch_weight_list = _validate_sub_epoch_summaries(self.constants, weight_proof)
+        if summaries is None:
+            log.warning("weight proof failed to validate sub epoch summaries")
+            return False, uint32(0)
+        return True, self.get_fork_point(summaries)
+
+    async def validate_weight_proof(
+        self, weight_proof: WeightProofV2, seed: bytes32, skip_segments: bool = False
+    ) -> Tuple[bool, uint32, List[SubEpochSummary], List[BlockRecord]]:
+        valid, summaries, records = await validate_weight_proof_no_fork_point(
+            self.constants, weight_proof, seed, skip_segments
+        )
+        return valid, self.get_fork_point(summaries), summaries, records
+
+    def get_fork_point(self, received_summaries: List[SubEpochSummary]) -> uint32:
+        # iterate through sub epoch summaries to find fork point
+        fork_point_index = 0
+        ses_heights = self.blockchain.get_ses_heights()
+        for idx, summary_height in enumerate(ses_heights):
+            log.debug(f"check summary {idx} height {summary_height}")
+            local_ses = self.blockchain.get_ses(summary_height)
+            if local_ses is None or local_ses.get_hash() != received_summaries[idx].get_hash():
+                break
+            fork_point_index = idx
+
+        if fork_point_index > 2:
+            # Two summeries can have different blocks and still be identical
+            # This gets resolved after one full sub epoch
+            height = ses_heights[fork_point_index - 2]
+        else:
+            height = uint32(0)
+
+        return height
+
+    async def _create_proof_of_weight(self, tip: bytes32, seed: bytes32) -> Optional[WeightProofV2]:
```

Comment:
How long does this take for example for a node in mainnet?

Reply:
if the database is fully populated with the sub_epoch objects this should take about 4 sec' 
the old one takes 2 sec'. 


i tried optimizing this as best i can but currently out of ideas